File: inference/fhe_layers/par_lower_diag_transpose.cpp
/*
 * Copyright (c) 2025-2026 CipherFlow (Shenzhen) Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include "par_lower_diag_transpose.h"
#include "layer_util.h"
#include <algorithm>
#include <cassert>
#include <stdexcept>
#include <utility>

using namespace std;
using namespace lattisense;

namespace {

bool is_power_of_two(uint32_t x) {
    return x > 0 && (x & (x - 1)) == 0;
}

}  // namespace

ParLowerDiagTranspose::ParLowerDiagTranspose(const CkksParameter& param_in,
                                             const Duo& shape,
                                             uint32_t n_heads,
                                             uint32_t head_dim,
                                             uint32_t level)
    : Layer(param_in) {
    assert(level >= 1);
    level_ = level;
    shape_ = shape;
    H_prepad_ = n_heads;
    m_ = head_dim;
    n_prepad_ = shape_[1];

    assert(H_prepad_ > 0);
    assert(m_ > 0 && is_power_of_two(m_));
    assert(shape_[0] == m_);
    assert(n_prepad_ > 0);

    H_ = next_pow2(H_prepad_);
    n_ = next_pow2(n_prepad_);
    assert(n_ % m_ == 0);

    n_slot_ = param_.get_n() / 2;
    segment_len_ = H_ * n_;
    assert(segment_len_ > 0);
    assert(n_slot_ % segment_len_ == 0);
    c_ = n_slot_ / segment_len_;
    assert(c_ > 0 && is_power_of_two(c_));
    assert(m_ % c_ == 0);
    m_c_ = m_ / c_;
}

uint32_t ParLowerDiagTranspose::expected_ct_count() const {
    return m_c_;
}

std::vector<double> ParLowerDiagTranspose::build_transpose_mask(uint32_t out_diag_idx, uint32_t mask_idx) const {
    assert(out_diag_idx < m_);
    assert(mask_idx < 2);
    vector<double> mask(n_slot_, 0.0);
    uint32_t out_local_idx = out_diag_idx % c_;
    uint32_t base = out_local_idx * segment_len_;

    if (mask_idx == 0) {
        for (uint32_t t = 0; t < n_ - out_diag_idx; t++) {
            uint32_t start = base + t * H_;
            for (uint32_t h = 0; h < H_; h++) {
                mask[start + h] = 1.0;
            }
        }
    } else {
        for (uint32_t t = 0; t < out_diag_idx; t++) {
            uint32_t start = (base + n_slot_ - out_diag_idx * H_ + t * H_) % n_slot_;
            for (uint32_t h = 0; h < H_; h++) {
                mask[start + h] = 1.0;
            }
        }
    }
    return mask;
}

void ParLowerDiagTranspose::prepare_weight() {
    CkksContext ctx = CkksContext::create_empty_context(param_);
    double mask_scale = param_.get_q(level_);

    transpose_mask_pt_.clear();
    transpose_mask_pt_.resize(m_);
    for (uint32_t out_diag_idx = 0; out_diag_idx < m_; out_diag_idx++) {
        transpose_mask_pt_[out_diag_idx].resize(2);
        for (uint32_t mask_idx = 0; mask_idx < 2; mask_idx++) {
            transpose_mask_pt_[out_diag_idx][mask_idx] =
                ctx.encode_ringt(build_transpose_mask(out_diag_idx, mask_idx), mask_scale);
        }
    }
}

CkksCiphertext ParLowerDiagTranspose::apply_mask(CkksContext& ctx,
                                                 const CkksCiphertext& input_level_l,
                                                 const CkksPlaintextRingt& mask_pt) const {
    double default_scale = param_.get_default_scale();
    auto mask_mul = ctx.ringt_to_mul(mask_pt, level_);
    return ctx.rescale(ctx.mult_plain_mul(input_level_l, mask_mul), default_scale);
}

std::vector<CkksCiphertext> ParLowerDiagTranspose::run_core(CkksContext& ctx,
                                                            const std::vector<CkksCiphertext>& input_cts) const {
    assert(input_cts.size() == m_c_);
    assert(transpose_mask_pt_.size() == m_);

    vector<CkksCiphertext> ct_ell_0(m_c_);
    vector<CkksCiphertext> ct_ell_1(m_c_);
    vector<bool> init_0(m_c_, false);
    vector<bool> init_1(m_c_, false);

    for (uint32_t j = 0; j < m_c_; j++) {
        for (uint32_t k = 0; k < c_; k++) {
            uint32_t source_diag_idx = c_ * j + k;
            uint32_t out_diag_idx = (m_ - (source_diag_idx % m_)) % m_;
            uint32_t ell = out_diag_idx / c_;
            uint32_t out_local_idx = out_diag_idx % c_;
            int rot = ((int)k - (int)out_local_idx) * (int)segment_len_ + (int)out_diag_idx * (int)H_;
            CkksCiphertext ct_rot = (rot == 0) ? input_cts[j].copy() : ctx.rotate(input_cts[j], rot);

            auto term0 = apply_mask(ctx, ct_rot, transpose_mask_pt_[out_diag_idx][0]);
            if (!init_0[ell]) {
                ct_ell_0[ell] = std::move(term0);
                init_0[ell] = true;
            } else {
                ct_ell_0[ell] = ctx.add(ct_ell_0[ell], term0);
            }

            auto term1 = apply_mask(ctx, ct_rot, transpose_mask_pt_[out_diag_idx][1]);
            if (!init_1[ell]) {
                ct_ell_1[ell] = std::move(term1);
                init_1[ell] = true;
            } else {
                ct_ell_1[ell] = ctx.add(ct_ell_1[ell], term1);
            }
        }
    }

    vector<CkksCiphertext> result;
    result.reserve(m_c_);
    for (uint32_t ell = 0; ell < m_c_; ell++) {
        assert(init_0[ell]);
        CkksCiphertext ct = std::move(ct_ell_0[ell]);
        if (init_1[ell]) {
            ct = ctx.add(ct, ctx.rotate(ct_ell_1[ell], -(int)segment_len_));
        }
        result.push_back(std::move(ct));
    }
    return result;
}

FeatureMatEncrypted ParLowerDiagTranspose::run(CkksContext& ctx, const FeatureMatEncrypted& input) {
    assert(input.level == level_);
    assert(input.head_shape[0] == m_ && input.head_shape[1] == n_prepad_);
    assert(input.shape[0] == H_prepad_ * input.head_shape[0] && input.shape[1] == input.head_shape[1]);
    assert(input.matmul_block_size == m_);
    assert(input.data.size() == expected_ct_count());

    FeatureMatEncrypted result(&ctx, input.level);
    Duo output_head_shape = {n_prepad_, m_};
    result.level = input.level - 1;
    result.shape = {output_head_shape[0], H_prepad_ * output_head_shape[1]};
    result.head_shape = output_head_shape;
    result.matmul_block_size = m_;
    result.data = run_core(ctx, input.data);
    return result;
}

Array<double, 2> ParLowerDiagTranspose::run_plaintext(const Array<double, 2>& A) const {
    assert(A.get_shape()[0] == H_prepad_ * m_);
    assert(A.get_shape()[1] == n_prepad_);

    Array<double, 2> T({n_prepad_, H_prepad_ * m_});
    for (uint32_t h = 0; h < H_prepad_; h++) {
        for (uint32_t i = 0; i < m_; i++) {
            for (uint32_t j = 0; j < n_prepad_; j++) {
                T.set(j, h * m_ + i, A.get(h * m_ + i, j));
            }
        }
    }
    return T;
}

File: inference/model_generator/layers/par_lower_diag_transpose.py
# Copyright (c) 2025-2026 CipherFlow (Shenzhen) Co., Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

import sys
from pathlib import Path
from collections import defaultdict

sys.path.insert(0, str(Path(__file__).resolve().parents[3]))

from inference.lattisense.frontend.custom_task import *
from inference.model_generator.layers.fhe_op_utils import memory_from_pt_counts


op_class = 'ParLowerDiagTranspose'


def _next_pow2(n: int) -> int:
    p = 1
    while p < n:
        p <<= 1
    return p


class ParLowerDiagTranspose:
    """Python model-generator counterpart of C++ ParLowerDiagTranspose."""

    def __init__(self, shape: tuple, n_heads: int, head_dim: int, n_slot: int):
        self.shape = shape
        self.H_prepad = n_heads
        self.m = head_dim
        self.n_prepad = shape[1]
        assert self.H_prepad > 0
        assert self.m > 0 and (self.m & (self.m - 1)) == 0
        assert shape[0] == self.m
        assert self.n_prepad > 0

        self.H = _next_pow2(self.H_prepad)
        self.n = _next_pow2(self.n_prepad)
        assert self.n % self.m == 0

        self.n_slot = n_slot
        self.segment_len = self.H * self.n
        assert self.segment_len > 0
        assert n_slot % self.segment_len == 0
        self.c = n_slot // self.segment_len
        assert self.c > 0 and (self.c & (self.c - 1)) == 0
        assert self.m % self.c == 0
        self.m_c = self.m // self.c

    def _apply_mask(self, ct, mask_pt):
        return rescale(mult(ct, mask_pt))

    def _run_core(self, input_cts: list, transpose_mask_pt: list) -> list:
        assert len(input_cts) == self.m_c

        ct_ell_0 = [None] * self.m_c
        ct_ell_1 = [None] * self.m_c

        for j in range(self.m_c):
            for k in range(self.c):
                source_diag_idx = self.c * j + k
                out_diag_idx = (self.m - (source_diag_idx % self.m)) % self.m
                ell = out_diag_idx // self.c
                out_local_idx = out_diag_idx % self.c
                rot = (k - out_local_idx) * self.segment_len + out_diag_idx * self.H
                ct_rot = input_cts[j] if rot == 0 else rotate_cols(input_cts[j], [rot])[0]

                term0 = self._apply_mask(ct_rot, transpose_mask_pt[out_diag_idx][0])
                ct_ell_0[ell] = term0 if ct_ell_0[ell] is None else add(ct_ell_0[ell], term0)

                term1 = self._apply_mask(ct_rot, transpose_mask_pt[out_diag_idx][1])
                ct_ell_1[ell] = term1 if ct_ell_1[ell] is None else add(ct_ell_1[ell], term1)

        result = []
        for ell in range(self.m_c):
            assert ct_ell_0[ell] is not None
            ct = ct_ell_0[ell]
            if ct_ell_1[ell] is not None:
                ct = add(ct, rotate_cols(ct_ell_1[ell], [-self.segment_len])[0])
            result.append(ct)
        return result

    def call(self, input_cts: list, transpose_mask_pt: list) -> list:
        return self._run_core(input_cts, transpose_mask_pt)

    def call_custom_compute(self, input_cts: list, data_source) -> list:
        transpose_mask_pt = []
        for out_diag_idx in range(self.m):
            row = []
            for mask_idx in range(2):
                node = CkksPlaintextRingtNode(f'encode_pt_transpose_mask_{out_diag_idx}_{mask_idx}')
                custom_compute(
                    inputs=[data_source],
                    output=node,
                    type='encode_pt',
                    attributes={
                        'op_class': op_class,
                        'type': 'transpose_mask_pt',
                        'out_diag_idx': out_diag_idx,
                        'mask_idx': mask_idx,
                    },
                )
                row.append(node)
            transpose_mask_pt.append(row)
        return self.call(input_cts, transpose_mask_pt)

    def get_memory(self, bytes_per_plaintext: int = 0) -> dict[str, int]:
        counts = {'weight': 0, 'bias': 0, 'mask': 2 * self.m}
        return memory_from_pt_counts(counts, bytes_per_plaintext)

    def get_fhe_op_count(self, level: int) -> dict:
        """Count FHE primitive operations grouped by level."""
        ops = defaultdict(lambda: {'rotate': 0, 'mult_plain': 0, 'mult': 0, 'add': 0, 'rescale': 0})
        lv = level

        seen_per_ell = [0] * self.m_c
        for j in range(self.m_c):
            for k in range(self.c):
                source_diag_idx = self.c * j + k
                out_diag_idx = (self.m - (source_diag_idx % self.m)) % self.m
                ell = out_diag_idx // self.c
                out_local_idx = out_diag_idx % self.c
                rot = (k - out_local_idx) * self.segment_len + out_diag_idx * self.H
                if rot != 0:
                    ops[lv]['rotate'] += 1

                # Two masks are applied for each source diagonal.
                ops[lv]['mult_plain'] += 2
                ops[lv]['rescale'] += 2

                # Accumulation into ct_ell_0/ct_ell_1 after the first term for each ell.
                if seen_per_ell[ell] > 0:
                    ops[lv - 1]['add'] += 2
                seen_per_ell[ell] += 1

        # Combine the two ell streams; ct_ell_1 is always present for each ell.
        ops[lv - 1]['rotate'] += self.m_c
        ops[lv - 1]['add'] += self.m_c

        return dict(ops)
